fix(parsestatus): add response data to error

ismay opened the pull request:
As discussed on [app-platform](https://dhis2.slack.com/archives/CLJSVA1RV/p1586945462166900). The body of the response that was attached to the error was already parsed, so once it reaches the app, it can't be parsed a second time and the data was lost.

This attaches the parsed data to the error under error.details (it if can be parsed). So error.details will now no longer be the response object.

p.s.: I used a try catch block as I found the synchronous notation easier to read in combination with async/await, but that's a bit subjective, so let me know if you'd like me to use the promise api.

amcgee commented on services/data/src/links/RestAPILink/fetchData.ts:
```diff
@@ -15,26 +15,36 @@ export const parseStatus = async (response: Response) => {
-                return response.status === 401 ? 'Unauthorized' : 'Forbidden'
-            })
+        let message = ''
+        let details
```
I wonder... should `details` be an empty object if the JSON is bad?  That way `error.details` will always be an object, never undefined.

ismay replied:
So I've pushed this 5ff6393, which fails on CI at the moment. I've set the type for details to an object, but since it's reassigned and then expected to contain a message the typescript compiler complains. What would be the best way to address that?

ismay replied:
See [the type definition](https://github.com/dhis2/app-runtime/pull/510/files#diff-b90ef08b9e29525a5ae5c77e44bd370dR2) and one of the fallbacks [here](https://github.com/dhis2/app-runtime/pull/510/files#diff-2f6d8eb1e7f7ef53661593022e859b46R19)

ismay replied:
Ok, pushed some changes that I think address this properly, but let me know if there are better ways to handle the typing here.

amcgee commented on services/data/src/links/RestAPILink/fetchData.ts:
```diff
@@ -15,26 +15,36 @@ export const parseStatus = async (response: Response) => {
     }
+
     if (response.status < 200 || response.status >= 400) {
+        let details
```
Similarly here

amcgee commented on services/data/src/links/RestAPILink/fetchData.ts:
```diff
@@ -15,26 +15,36 @@ export const parseStatus = async (response: Response) => {
+        let details
+
+        try {
+            details = await response.json()
```
Does `response.json()` work for primitive JSON types like a string?  I imagine not, so I think this should always be an object or throw...

amcgee replied:
Actually [looks like it could be anything](https://developer.mozilla.org/en-US/docs/Web/API/Body/json), not just an Object.

Sorry to complicate this... possibly for a later fix... but maybe if we get just a string back in the body we should use that as the message?  I think I've seen that happen in a few rare cases in the DHIS2 API...

ismay replied:
No problem. Yeah there are a couple of assumptions made. We're assuming that the response can be parsed as json, and expecting that it'll parse to an object.

The first is dealt with with the try catch, and then we could check whether it's a string or an object and deal with those cases. And then we'd ignore number and array (as in, just assign those to details). How would you feel about that?

ismay replied:
I went ahead and implemented the above suggestion, let me know what you think.

amcgee replied:
Sorry, apparently I the mozilla docs page was misleading... commented

varl commented on services/data/src/links/RestAPILink/fetchData.ts:
```diff
@@ -15,26 +15,36 @@ export const parseStatus = async (response: Response) => {
-            .catch(() => {
-                return response.status === 401 ? 'Unauthorized' : 'Forbidden'
-            })
+        let message = ''
+        let details
+
+        try {
+            details = await response.json()
+            message = details.message
+        } catch (e) {
+            message = response.status === 401 ? 'Unauthorized' : 'Forbidden'
+        }
```
I prefer the former; it's a simpler syntax and I prefer that the `catch` is intentionally used and not a dummy statement.

amcgee replied:
Yeah it looked simpler in my head I think...

ismay replied:
@amcgee I've removed the default for message: 5ebd4c9. Let me know what you think.

ismay commented:
> I suppose we'll need to make sure this is flagged as a BREAKING CHANGE when merging too, though it's a silly one. Perhaps we could call this a "fix" since it was never specified in the docs and accessing the body of the response never worked?

Yeah a fix is fine with me. It all depends on what the intended api was I guess. So you'd prefer fix?

ismay commented:
I think I've addressed all the comments with the above commits. Let me know how you feel about the current state of the PR, and whether you'd prefer a fix or a breaking change, and we can add some final tests and this should be good to go!

amcgee commented:
> Yeah a fix is fine with me. It all depends on what the intended api was I guess. So you'd prefer fix?

Yeah, it was basically broken before so let's say this is a fix.

amcgee commented on services/data/src/links/RestAPILink/fetchData.ts:
```diff
@@ -10,33 +10,66 @@ export const parseContentType = (contentType: string | null) => {
+    const accessError =
         response.status === 401 ||
         response.status === 403 ||
+        response.status === 404 ||
```
I worry about adding this as an `access` type error, since we lose the ability to differentiate between Not Found and Not Allowed if we lump them together and don't pass through a status code.

Can we leave the `404` check out of this PR and attack that separately?

ismay replied:
Yeah I agree, I'll remove it 👍

amcgee commented on services/data/src/links/RestAPILink/fetchData.ts:
```diff
@@ -10,33 +10,66 @@ export const parseContentType = (contentType: string | null) => {
+        try {
+            details = await response.json()
+
+            if (typeof details === 'string') {
+                message = details
+
+                // Clearing details, since it's expected to be an object and not a string
+                details = {}
+            } else if (typeof details === 'object') {
+                message = details.message
+            }
```
So I did a quick test in Chrome console and it looks like the "raw text" option here is a bit more nuanced:

```js
> (new Response("This is a test")).json().then(console.log)
VM1441:1 Uncaught (in promise) SyntaxError: Unexpected token T in JSON at position 0
> (new Response('"This is a test"')).json().then(console.log)
This is a test
```

So I think the "string" case we're catching here (wrapped in double quotes) will actually be super rare.

I think we should actually keep this as-is, just try to parse the body as JSON, assume it's an object, and use `details.message` if it exists.  So something like this:

```js
try {
    details = await response.json()
    message = details.message
} catch (e) {
    // Do nothing
}
// Moving this outside the catch means we put always have something in `message` even if the returned JSON doesn't have a `message` property
if (!message) {
    message = response.status === 401 ? 'Unauthorized' : 'Forbidden'
}
```

---
I initially took a stab at a more "exhaustive" option, which will also accept non-JSON bodies and use make a custom details object out of them.  I'm not sure I like it as much... it is liberal in what it accepts from the server, but maybe in this case since we control the server we should be more conservative and fix the unexpected responses before the server sends them?

```js
const errorText = await response.text()
try {
    details = await JSON.parse(errorText) // I think we can assume this is an object here...
    message = details.message
} catch (e) {
    message = errorText
}
if (!message || !message.length) {
    message = response.status === 401 ? 'Unauthorized' : 'Forbidden'
}
```

Let me know which you prefer!

ismay replied:
I'd prefer the former. I think it's better to assume uniformity across our server responses. If we don't have that, it'd be good to have it fail in my opinion instead of smoothing it over.

ismay commented:
> Thanks @ismay !! Sorry for the runaround on such a small change, added a couple more comments. I think we should leave the 404 addition for another change.

No problem, I think it's good to think these kinds of changes over. I think I've addressed all the comments and I've removed the breaking change label. If you're ok with it let me know, I propose before merge we squash to the first commit message.

amcgee commented on services/data/src/links/RestAPILink/fetchData.ts:
```diff
@@ -10,33 +10,51 @@ export const parseContentType = (contentType: string | null) => {
+
+    if (accessError) {
+        let message
+        let details = {} as any
```
I think I saw you mention this in a comment, sorry I missed it - the issue here I assume is that `details` is assigned the type of an **empty** object with `let details = {}`, and so typescript complains about dereferencing `details.message`, is that right?

If so, I think the right thing to do is to change the FetchErrorDetails type to be what we actually expect it to be - an Object with an optional `message` prop.  So that would be 

Once the type is changed we can actually say `let details: FetchErrorDetails = {}` 

I tested this quickly [here](https://www.typescriptlang.org/play/?ssl=1&ssc=47&pln=1&pc=1#code/C4TwDgpgBAYhwGMAWBRATmg9mgIvAhgJYA2AzlALxQDeUAthKafgObQD8AXFKcGoQDsWUAL4AoYvCgATAiVLc4iVBmx5gRMpRrixsjfO20GTVhG4ByAB4gAXhdFR85fAJBiEmAaUySAdMSYLAAU+pqkfibMbACUQA), seems to work and has the added benefit of being extra explicit about what we expect to be in the `details` object.

ismay replied:
Ah ok, thanks. I've updated it to match your suggestions 👍

amcgee commented on services/data/src/engine/types/FetchError.ts:
```diff
@@ -1,17 +1,17 @@
 export type FetchErrorTypeName = 'network' | 'unknown' | 'access' | 'aborted'
-export type FetchErrorDetails = Response | Error
+export type FetchErrorDetails = object
```
Following along from the comment above, I think we should be super-explicit here and use `export type FetchErrorDetails = { message?: string }`

amcgee replied:
This will also allow someone catching a `FetchError` object to know (statically or with IDE auto-completion) that there is a possibly-undefined string property `message` on the `details` object, which is correct 👍

ismay replied:
See https://github.com/dhis2/app-runtime/pull/510#discussion_r412846708

ismay commented:
All right, done and done. Thanks for the help @amcgee 👍!
